junit5/src/main/java/org/jboss/weld/junit5/WeldInitiator.java

```diff
@@ -0,0 +1,172 @@
+
+import java.lang.annotation.Annotation;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.spi.Bean;
+
+import org.jboss.weld.environment.se.Weld;
+import org.jboss.weld.environment.se.WeldContainer;
+import org.jboss.weld.junit.common.AbstractWeldInitiator;
+
+/**
+ * TODO document this
+ *
+ * @author Matej Novotny
+ */
+public class WeldInitiator extends AbstractWeldInitiator {
+
+    /**
+     * The container is configured with the result of {@link #createWeld()} method and the given bean classes are added.
+     *
+     * @param beanClasses
+     * @return a new test rule
+     * @see Weld#beanClasses(Class...)
+     */
+    public static WeldInitiator of(Class<?>... beanClasses) {
+        return of(createWeld().beanClasses(beanClasses));
+    }
+
+    /**
+     * The container is configured through a provided {@link Weld} instance.
+     *
+     * @param weld
+     * @return a new test rule
+     */
+    public static WeldInitiator of(Weld weld) {
+        return from(weld).build();
+    }
+
+    /**
+     * The container is configured with the result of {@link #createWeld()} method and all the classes from the test class
+     * package are added.
+     *
+     * @return a new test rule
+     */
+    public static WeldInitiator ofTestPackage() {
+        return fromTestPackage().build();
+    }
+
+    /**
+     * Create a builder instance.
+     *
+     * @param weld
+     * @return a builder instance
+     * @see #of(Class...)
+     */
+    public static Builder from(Class<?>... beanClasses) {
+        return from(createWeld().beanClasses(beanClasses));
+    }
+
+    /**
+     * Create a builder instance.
+     *
+     * @param weld
+     * @return a builder instance
+     * @see #of(Weld)
+     */
+    public static Builder from(Weld weld) {
+        return new Builder(weld);
+    }
+
+    /**
+     * Create a builder instance.
+     *
+     * @return a builder instance
+     * @see #ofTestPackage()
+     */
+    public static Builder fromTestPackage() {
+        return new Builder(null);
+    }
+
+    /**
+     * This builder can be used to customize the final {@link WeldInitiator} instance, e.g. to activate a context for a given
+     * normal scope.
+     */
+    public static final class Builder extends AbstractBuilder {
+
+        public Builder(Weld weld) {
+            super(weld);
+        }
+
+        /**
+         * Activate and deactivate contexts for the given normal scopes for each test method execution.
+         * <p>
+         * {@link ApplicationScoped} is ignored as it is always active.
+         * </p>
+         *
+         * @param normalScopes
+         * @return self
+         */
+        @SafeVarargs
+        public final Builder activate(Class<? extends Annotation>... normalScopes) {
+            return (Builder) super.activate(normalScopes);
```

Comment:
Here we can instead do:
`super.activate(normalScopes);`
`return this;`

But we still need it just to return `Builder` type (and not `AbstractBuilder`) - same happens in junit 4 WeldInitiator.

Reply:
There is a trick to handle hierarchy in builders... see for example https://github.com/cdi-spec/cdi-tck/blob/master/impl/src/main/java/org/jboss/cdi/tck/shrinkwrap/ArchiveBuilder.java#L93. In this case something like `AbstractBuilder<T extends AbstractBuilder<T>>` and add abstract method `T self()` which should be used in `AbstractBuilder`, i.e. `return self()` instead of `return this`.